From 40e306eac33466fa27a452fae179126510d299db Mon Sep 17 00:00:00 2001
From: Matija Tudan <matija.tudan@nulix.hr>
Date: Mon, 17 Nov 2025 19:36:45 +0000
Subject: [PATCH 1/1] mmc: rockchip_dw_mmc: Always apply default sample phase
 during probe

Recent changes modified the logic that determines when clk_set_phase()
should be called for the MMC sample clock:

    else if ((!priv->sample_clk.dev))
        ret = clk_set_phase(...);

This caused boards where priv->sample_clk.dev is non-NULL (but does not
reflect a real clock provider) to skip sample phase configuration
entirely, resulting in MMC/SD/eMMC failing to initialize in U-Boot
properly.

Add a fallback path to always call clk_set_phase() when the initial
condition is not met. This restores the previous behavior and fixes
MMC detection on platforms where the sample clock is internal to the
controller and not represented as a standalone clk device.

This avoids reverting the upstream commit while ensuring correct
operation on boards requiring explicit sample phase programming.

Signed-off-by: Matija Tudan <matija.tudan@nulix.hr>
---
 drivers/mmc/rockchip_dw_mmc.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/mmc/rockchip_dw_mmc.c b/drivers/mmc/rockchip_dw_mmc.c
index 4e5ab3d4935..d1bf6e2c11f 100644
--- a/drivers/mmc/rockchip_dw_mmc.c
+++ b/drivers/mmc/rockchip_dw_mmc.c
@@ -449,6 +449,8 @@ internal_phase:
 			ret = rockchip_mmc_set_phase(host, true, plat->mmc.default_phase);
 		else if ((!priv->sample_clk.dev))
 			ret = clk_set_phase(&priv->sample_clk, plat->mmc.default_phase);
+		else
+			ret = clk_set_phase(&priv->sample_clk, plat->mmc.default_phase);
 		if (ret < 0)
 			debug("MMC: can not set default phase!\n");
 	}
-- 
2.43.7

